### PR TITLE
Allow exporting attachment values and types from scripts

### LINKED ISSUE #2226: Attachment.getType() results in error
### Problem

**System info:** Running flow-cli pre-release version `v0.43.0-cadence-attachments-2` on macOS Ventura 13.1

I defined a basic attachment in a simple contract. When running a script trying to call `getType()` on the static type of that attachment, I encountered an error with the following stack trace:

```
WARN[0006] ❗  Script reverted                            computationUsed=0 scriptID=168dfdbe77dea60a1323ef9fe6a44df4884f00929df58169f4cfcb606e34da4a
WARN[0006] ERR [168dfd] [Error Code: 1101] error caused by: [Error Code: 1101] cadence runtime error: Execution failed:
error: internal error: cannot export composite type MyAttachment of unknown kind CompositeKindAttachment
goroutine 6384 [running]:
runtime/debug.Stack()
	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
github.com/onflow/cadence/runtime/errors.NewUnexpectedError({0x1030f5e39?, 0x103816780?}, {0x1400adf0db8?, 0x10202845c?, 0x103a8a260?})
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/errors/errors.go:125 +0x3c
github.com/onflow/cadence/runtime.getWrappedError({0x103816780?, 0x1400af5c710?}, {0x103b7d500, 0x14005e1ca20}, {0x14006891b00?, 0x14006891b30?})
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/runtime.go:254 +0x318
github.com/onflow/cadence/runtime.(*interpreterRuntime).Recover(0x140007ab520?, 0x1400adf1ba0, {0x103b7d500, 0x14005e1ca20}, {0x14006891b00?, 0x14006891b30?})
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/runtime.go:225 +0x5c
panic({0x103816780, 0x1400af5c710})
	/usr/local/go/src/runtime/panic.go:884 +0x204
github.com/onflow/cadence/runtime.exportCompositeType({0x103b66b80, 0x14009b095c0}, 0x1400a76f540, 0x0?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:433 +0x498
github.com/onflow/cadence/runtime.ExportMeteredType.func1({0x103b931c0?, 0x1400a76f540?}, {0x103b66b80?, 0x14009b095c0?}, 0x1400adf13c8?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:208 +0x350
github.com/onflow/cadence/runtime.ExportMeteredType({0x103b66b80, 0x14009b095c0}, {0x103b931c0, 0x1400a76f540}, 0x1400adf1438?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:329 +0x90
github.com/onflow/cadence/runtime.exportFunctionType({0x103b66b80, 0x14009b095c0}, 0x14008a7b260, 0x0?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:561 +0xa8
github.com/onflow/cadence/runtime.ExportMeteredType.func1({0x103b93440?, 0x14008a7b260?}, {0x103b66b80?, 0x14009b095c0?}, 0x1400adf15a8?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:214 +0x1ac
github.com/onflow/cadence/runtime.ExportMeteredType({0x103b66b80, 0x14009b095c0}, {0x103b93440, 0x14008a7b260}, 0x1400adf1658?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertTypes.go:329 +0x90
github.com/onflow/cadence/runtime.exportTypeValue({{0x103b85290?, 0x14008a7b260?}}, 0x3?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertValues.go:588 +0xb0
github.com/onflow/cadence/runtime.exportValueWithInterpreter({0x103b90d28?, 0x1400af5c6e0?}, 0x14009b095c0, {{0x0?, 0x0?}, {0x0?, 0x0?}}, 0x1400adf1c40)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertValues.go:219 +0x1350
github.com/onflow/cadence/runtime.exportValue(...)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/convertValues.go:40
github.com/onflow/cadence/runtime.(*interpreterScriptExecutor).execute(0x140003a6900)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/script_executor.go:195 +0x1b8
github.com/onflow/cadence/runtime.(*interpreterScriptExecutor).Execute.func1()
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/script_executor.go:73 +0x24
sync.(*Once).doSlow(0x0?, 0x2?)
	/usr/local/go/src/sync/once.go:74 +0x104
sync.(*Once).Do(...)
	/usr/local/go/src/sync/once.go:65
github.com/onflow/cadence/runtime.(*interpreterScriptExecutor).Execute(0x140003a6900?)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/script_executor.go:72 +0x50
github.com/onflow/cadence/runtime.(*interpreterScriptExecutor).Result(0x140003a6900)
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/script_executor.go:82 +0x20
github.com/onflow/cadence/runtime.(*interpreterRuntime).ExecuteScript(0x140068f12e0, {{0x1400ab29400, 0x2143, 0x2500}, {0x0, 0x0, 0x0}}, {{0x103b9e628, 0x14007d629a0}, {0x103b7d500, ...}, ...})
	/go/pkg/mod/github.com/onflow/cadence@v0.30.0-attachments/runtime/runtime.go:266 +0x18c
github.com/onflow/flow-go/fvm/runtime.WrappedCadenceRuntime.ExecuteScript({{0x103b8da58?, 0x140068f12e0?}}, {{0x1400ab29400, 0x2143, 0x2500}, {0x0, 0x0, 0x0}}, {{0x103b9e628, 0x14007d629a0}, ...})
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/runtime/wrapped_cadence_runtime.go:26 +0x9c
github.com/onflow/flow-go/fvm/runtime.(*ReusableCadenceRuntime).ExecuteScript(...)
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/runtime/reusable_cadence_runtime.go:157
github.com/onflow/flow-go/fvm.(*scriptExecutor).execute(0x14005e225a0)
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/script.go:196 +0x2d8
github.com/onflow/flow-go/fvm.(*scriptExecutor).Execute(0x14005e225a0)
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/script.go:158 +0x20
github.com/onflow/flow-go/fvm.Run({0x103b77fa0?, 0x14005e225a0})
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/fvm.go:39 +0xac
github.com/onflow/flow-go/fvm.(*VirtualMachine).Run(_, {0x0, 0x186a0, 0xffffffffffffffff, 0x3e80, 0xf424000, 0x1312d00, {0x1, 0x1, 0x3e8, ...}, ...}, ...)
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/fvm/fvm.go:120 +0x754
github.com/onflow/flow-emulator.(*Blockchain).ExecuteScriptAtBlock(0x1400044c000, {0x1400ab29400, 0x2143, 0x2500}, {0x0, 0x0, 0x0}, 0x1bb?)
	/go/pkg/mod/github.com/onflow/flow-emulator@v0.41.0-cadence-attachments-2/blockchain.go:1176 +0x37c
github.com/onflow/flow-emulator/server/backend.(*Backend).executeScriptAtBlock(0x140086bc720, {0x1400ab29400?, 0x103137a58?, 0x103b782a0?}, {0x0?, 0x1400af6e3f0?, 0x0?}, 0x1400af6e420?)
	/go/pkg/mod/github.com/onflow/flow-emulator@v0.41.0-cadence-attachments-2/server/backend/backend.go:604 +0x40
github.com/onflow/flow-emulator/server/backend.(*Backend).ExecuteScriptAtLatestBlock(0x140086bc720, {0x102026244?, 0x2?}, {0x1400ab29400, 0x2143, 0x2500}, {0x0, 0x0, 0x0})
	/go/pkg/mod/github.com/onflow/flow-emulator@v0.41.0-cadence-attachments-2/server/backend/backend.go:411 +0x1a4
github.com/onflow/flow-emulator/server/backend.(*Adapter).ExecuteScriptAtLatestBlock(0x1039757e0?, {0x103b7bdd0?, 0x14009c93fb0?}, {0x1400ab29400?, 0x140090a9908?, 0x102fa5c90?}, {0x0?, 0x102a41dd8?, 0x103ab1140?})
	/go/pkg/mod/github.com/onflow/flow-emulator@v0.41.0-cadence-attachments-2/server/backend/adapter.go:161 +0x30
github.com/onflow/flow-go/access.(*Handler).ExecuteScriptAtLatestBlock(0x140090a9938?, {0x103b7bdd0?, 0x14009c93fb0?}, 0x1400af6e420?)
	/go/pkg/mod/github.com/onflow/flow-go@v0.28.1-0.20230104183005-ff76d50986c7/access/handler.go:361 +0x88
github.com/onflow/flow/protobuf/go/flow/access._AccessAPI_ExecuteScriptAtLatestBlock_Handler.func1({0x103b7bdd0, 0x14009c93fb0}, {0x103a81a60?, 0x140068ef1d0})
	/go/pkg/mod/github.com/onflow/flow/protobuf/go/flow@v0.3.2-0.20221130185733-92eb85ead310/access/access_grpc.pb.go:810 +0x74
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1({0x103b7bdd0, 0x14009c93fb0}, {0x103a81a60, 0x140068ef1d0}, 0x140000ef9f8?, 0x1400a8da6f0)
	/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x74
github.com/onflow/flow/protobuf/go/flow/access._AccessAPI_ExecuteScriptAtLatestBlock_Handler({0x103afd140?, 0x140086d0540}, {0x103b7bdd0, 0x14009c93fb0}, 0x14000554690, 0x14000022260)
	/go/pkg/mod/github.com/onflow/flow/protobuf/go/flow@v0.3.2-0.20221130185733-92eb85ead310/access/access_grpc.pb.go:812 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0x14008512000, {0x103b8bad8, 0x1400ad7e1a0}, 0x140069ade60, 0x140086d05d0, 0x104856a98, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1340 +0xb7c
google.golang.org/grpc.(*Server).handleStream(0x14008512000, {0x103b8bad8, 0x1400ad7e1a0}, 0x140069ade60, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:1713 +0x82c
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:965 +0x84
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.50.1/server.go:963 +0x290

--> 168dfdbe77dea60a1323ef9fe6a44df4884f00929df58169f4cfcb606e34da4a
```

### Steps to Reproduce

Here is the contract I deployed defining a base resource and an attachment:

```
pub contract AttachmentsTest {

    // Base resource
    pub resource Base {
        pub let id: UInt64
        pub let name: String

        init(name: String) {
            self.id = self.uuid
            self.name = name
        }

    // Basic attachment implementing the above interface designed for any 
    // resource implementing MyInterface
    pub attachment MyAttachment for Base {
        pub let id: UInt64
        pub let name: String
        init() {
            self.id = base.id
            self.name = base.name
        }

        pub fun getName(): String {
            return self.name
        }
    }
}

```

And the script I ran which resulted in the error

```
import AttachmentsTest from "../contracts/AttachmentsTest.cdc"

pub fun main(): Type {
    return AttachmentsTest.MyAttachment.getType()
}
```

As a frame of reference, this is what I get when running the above script, but on the `Base` resource static type

```
Result: Type<((String):A.f8d6e0586b0a20c7.AttachmentsTest.Base)>()
```

### PR DESCRIPTION
**Pull request by dsainati1**
Closes #2226
______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [X] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: values.go [hunk 1]
```diff
@@ -1645,6 +1645,63 @@ func (v Resource) String() string {
 	return formatComposite(v.ResourceType.ID(), v.ResourceType.Fields, v.Fields)
 }
 
+// Resource
```
Comment (SupunS) [inline values.go:1648]:
```suggestion
// Attachment
```

File: types.go [hunk 2]
```diff
@@ -1090,6 +1090,80 @@ func (t *ResourceType) CompositeInitializers() [][]Parameter {
...
 
+type AttachmentType struct {
+	Location            common.Location
+	BaseType            Type
+	QualifiedIdentifier string
+	Fields              []Field
+	Initializers        [][]Parameter
+}
+
+// AttachmentType
```
Comment (SupunS) [inline types.go:1101]: Guess this comment should move to the top of `AttachmentType` struct

**Review by SupunS — APPROVED** [review]

File: runtime/attachments_test.go [hunk 3]
```diff
@@ -147,6 +147,175 @@ func TestAccountAttachmentSaveAndLoad(t *testing.T) {
...
+			Source: deployTx,
+		},
+		Context{
+			Interface: runtimeInterface1,
+			Location:  nextTransactionLocation(),
+		},
+	)
+	require.NoError(t, err)
+
+	_, err = rt.ExecuteScript(
```
Comment (turbolent) [inline runtime/attachments_test.go:307]: Same here, maybe assert that the exported value matches the interpreter value.

File: runtime/tests/interpreter/attachments_test.go [hunk 4]
```diff
@@ -1645,3 +1645,27 @@ func TestInterpretAttachmentResourceReferenceInvalidation(t *testing.T) {
...
+            attachment A for R {}
+            fun test() {
+                let r <- create R()
+                let r2 <- attach A() to <-r
+                let a: Type = r2[A]!.getType()
+                destroy r2
+            }
+        `)
+
+		_, err := inter.Invoke("test")
```
Comment (turbolent) [inline runtime/tests/interpreter/attachments_test.go:1668]: Maybe return `a` from `test` and assert the returned run-time type's identifier matches the expected attachment type.

File: values.go [hunk 5]
```diff
@@ -1645,6 +1645,63 @@ func (v Resource) String() string {
...
+
+type Attachment struct {
+	AttachmentType *AttachmentType
+	Fields         []Value
+}
+
+var _ Value = Resource{}
```
Comment (turbolent) [inline values.go:1655]:
```suggestion
var _ Value = Attachment{}
```

**Review by turbolent — COMMENTED** [review]
Looks good!

Maybe also add tests to `convertValues_test.go`, i.e. `TestExportValue`.

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**